r2taint/taint.py

```diff
@@ -0,0 +1,346 @@
+
+
+parser = argparse.ArgumentParser( description='static taint analysis' )
+parser.add_argument("-reg", action="append", default=[], help='taint register')
+parser.add_argument("-mem", action="append", default=[], help='taint memory')
+parser.add_argument("-deep", type=int, default=1, help='max analysis deep')
+parser.add_argument("-v", type=bool, default=False, help='verbose')
+args = parser.parse_args()
+r2 = r2pipe.open()
+
+
+def get_sub_registers(register):
+	register = register.lower()
+	for sub_registers in [ 
+			['rax', 'eax', 'ax', 'ah', 'al'],
+			['rdx', 'edx', 'dx', 'dh', 'dl'],
+			['rcx', 'ecx', 'cx', 'ch', 'cl'],
+			['rbx', 'ebx', 'bx', 'bh', 'bl'],
+			['rbp', 'esp', 'sp'],
+			['rsp', 'ebp', 'bp'],
+			['rdi', 'edi', 'di'],
+			['rsi', 'esi', 'si'],
+			['r8', 'r8d', 'r8w', 'r8b'],
+			['r9', 'r9d', 'r9w', 'r9b'],
+			['r10', 'r10d', 'r10w', 'r10b'],
+			['r11', 'r11d', 'r11w', 'r11b'],
+			['r12', 'r12d', 'r12w', 'r12b'],
+			['r13', 'r13d', 'r13w', 'r13b'],
+			['r14', 'r14d', 'r14w', 'r14b'],
+			['r15', 'r15d', 'r15w', 'r15b'],
+		]:
+		if register in sub_registers:
+			index = sub_registers.index(register)
+			if index < 3:
+				return sub_registers[index:] # AX has [AH,AL]
+			else:
+				return [sub_registers[index]] # AH hasn't AL
+	return [register]
+
+def taint_blacklist(instr):
+	if instr["mnemonic"] in ("xor",):
+		if instr["opex"]["operands"][0]["type"] == "reg" and instr["opex"]["operands"][1]["type"] == "reg":
+			if instr["opex"]["operands"][0]["value"] == instr["opex"]["operands"][1]["value"]:
+				return True
+	else:
+		return False
+
+regs_taint = set()
+mems_taint = set()
+tainted = []
+def taint(rip, instr, access):
+	global regs_taint, mems_taint, tainted
+	((regs_r, regs_w), (mems_r, mems_w)) = access
+	is_propagation = False
+
+	for reg_r in regs_r:
+		if reg_r in regs_taint:
+			is_propagation = True
+	for mem_r in mems_r:
+		if mem_r in mems_taint:
+			is_propagation = True
+
+	if taint_blacklist(instr):
+		is_propagation = False
+
+	if is_propagation:
+		out = "[taint] {addr}: {instr}".format(addr=hex(rip), instr=instr["disasm"])
+		if not out in tainted:
+			tainted.append(out)
+		if args.v:
+			print out
+		r2.cmd("ecHi yellow @{addr}".format(addr=rip))
+		for reg_w in regs_w:
+			if instr["mnemonic"] in ('push','pop'):
+				if reg_w in ('rsp','esp','sp'):
+					continue
+			for reg in get_sub_registers(reg_w):
+				regs_taint.add(reg)
+		for mem_w in mems_w:
+			mems_taint.add(mem_w)
+	else:
+		for reg_w in regs_w:		
+			for reg in get_sub_registers(reg_w):
+				if reg in regs_taint:
+					regs_taint.remove(reg)
+		for mem_w in mems_w:
+			if mem_w in mems_taint:
+				mems_taint.remove(mem_w)
+				
+	return is_propagation
+
+def init_taint(args):
+	global regs_taint, mems_taint
+	regs_taint = set()
+	mems_taint = set()
+	ret = False
+	for reg in args.reg:
+		regs_taint.add(reg)
+		ret = True
+	for mem in args.mem:
+		try:
+			addr = int(mem, 16)
+			mems_taint.add(addr)
+			ret = True
+		except:
```

Comment:
shouldn't the exception being handled?

Reply:
Here is not need to handle. It is an attempt to interpret `-mem` as 0xADDR, else it will be interpret as flagname.